### ISSUE TITLE
Creative upscale use-case

### ISSUE THREAD

**Issue by cyrildiagne** [jy0205/Pyramid-Flow#149]
Hello,

Thank you for sharing this great work & codebase!

I was wondering if Pyramid-Flow could be a good match for the 4x video upscaling use-case.
It seems like maybe we could add noise to the low-resolution video/tile as replacement of the first block / stage, and then use the efficient pyramid generation pipeline to generate the high-resolution video/tile.

I toyed a bit with the idea but couldn't get satisfactory results with this quick test.
(I understand that the results would differ from the original video since we're adding noise to the original input)
But maybe I'm missing some fundamental insight?

I saw [this comment](https://github.com/jy0205/Pyramid-Flow/issues/96) about an hour-glass structure for video2video, but for the upscaling use-case, it seems like using the pyramid structure would be appropriate.

Any insights appreciated, thanks!

---

The input 96x96 low-res video:

https://github.com/user-attachments/assets/f1ba28ef-d50a-486d-9446-9e6cb32a749f

The 384x384 output, using the `pyramid_flux` model and prompt "FPV drone view of a medieval castle, details, high quality, Ultra HD, 4K, 8K"

https://github.com/user-attachments/assets/31b8a663-da09-4d6d-8224-972f2029b908

Surprisingly, if I override `sigma`to be 0.9 instead of `scheduler.end_sigmas[0]` in the `prepare_input_latents` function I get slightly better results, but still far off the original video:

https://github.com/user-attachments/assets/505496ba-7154-4fc2-af0e-395abe92f20e

```py
def upscale_one_unit(
    latents,
    past_conditions, # List of past conditions, contains the conditions of each stage
    prompt_embeds,
    prompt_attention_mask,
    pooled_prompt_embeds,
    num_inference_steps,
    height,
    width,
    temp,
    device,
    dtype,
    guidance_scale = 7,
    video_guidance_scale = 4,
    generator = None,
    is_first_frame: bool = False,
):
    intermed_latents = []

    # start from the 2nd stage since we already have noisy latents for the 1st stage
    for i_s in range(1, num_stages):
        scheduler.set_timesteps(num_inference_steps[i_s], i_s, device=device)
        timesteps = scheduler.timesteps

        if i_s > 0:
            height *= 2; width *= 2
            latents = rearrange(latents, 'b c t h w -> (b t) c h w')
            latents = F.interpolate(latents, size=(height, width), mode='nearest')
            latents = rearrange(latents, '(b t) c h w -> b c t h w', t=temp)

            # Fix the stage
            ori_sigma = 1 - scheduler.ori_start_sigmas[i_s]   # the original coeff of signal
            gamma = scheduler.config.gamma
            alpha = 1 / (math.sqrt(1 + (1 / gamma)) * (1 - ori_sigma) + ori_sigma)
            beta = alpha * (1 - ori_sigma) / math.sqrt(gamma)

            bs, ch, temp, height, width = latents.shape
            noise = sample_block_noise(bs, ch, temp, height, width)
            noise = noise.to(device=device, dtype=dtype)
            latents = alpha * latents + beta * noise    # To fix the block artifact

        for idx, t in enumerate(timesteps):
            do_classifier_free_guidance = guidance_scale > 1.0

            # expand the latents if we are doing classifier free guidance
            latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
        
            # broadcast to batch dimension in a way that's compatible with ONNX/Core ML
            timestep = t.expand(latent_model_input.shape[0]).to(latent_model_input.dtype)
            
            latent_model_input = past_conditions[i_s] + [latent_model_input]

            noise_pred = dit(
                sample=[latent_model_input],
                timestep_ratio=timestep,
                encoder_hidden_states=prompt_embeds,
                encoder_attention_mask=prompt_attention_mask,
                pooled_projections=pooled_prompt_embeds,
            )

            if noise_pred[0].isnan().any():
                raise ValueError(f"noise_pred")

            noise_pred = noise_pred[0]
            
            # perform guidance
            if do_classifier_free_guidance:
                noise_pred_uncond, noise_pred_text = noise_pred.chunk(2)
                if is_first_frame:
                    noise_pred = noise_pred_uncond + guidance_scale * (noise_pred_text - noise_pred_uncond)
                else:
                    noise_pred = noise_pred_uncond + video_guidance_scale * (noise_pred_text - noise_pred_uncond)
            
            # compute the previous noisy sample x_t -> x_t-1
            latents = scheduler.step(
                model_output=noise_pred,
                timestep=timestep,
                sample=latents,
                generator=generator,
            ).prev_sample

        intermed_latents.append(latents)

    return intermed_latents

def upscale(latents, prompt_embeds, prompt_attention_mask, pooled_prompt_embeds, num_inference_steps = [10, 10, 10], guidance_scale = 7):
    generated_latents_list = []

    # The video sequence: one frame + N * unit
    frame_per_unit = 1
    temp = latents.shape[2]
    num_units = 1 + (temp - 1) // frame_per_unit

    height = latents.shape[-2]
    width = latents.shape[-1]

    for unit_index in tqdm(range(num_units)):
        if unit_index == 0:
            past_condition_latents = [[] for _ in range(num_stages)]
            intermed_latents = upscale_one_unit(
                latents[:,:,:1],
                past_condition_latents,
                prompt_embeds,
                prompt_attention_mask,
                pooled_prompt_embeds,
                num_inference_steps,
                height,
                width,
                1,
                device,
                dtype,
                generator=generator,
                is_first_frame=True,
            )
        else:
            past_condition_latents = []
            clean_latents_list = get_pyramid_latent(
                torch.cat(generated_latents_list, dim=2), num_stages - 1
            )

            do_classifier_free_guidance = guidance_scale > 1.0
            for i_s in range(num_stages):
                last_cond_latent = clean_latents_list[i_s][:,:,-frame_per_unit:]
                stage_input = [torch.cat([last_cond_latent] * 2) if do_classifier_free_guidance else last_cond_latent]
                
                # pad the past clean latents
                cur_unit_num = unit_index
                cur_stage = i_s
                cur_unit_ptx = 1

                while cur_unit_ptx < cur_unit_num:
                    cur_stage = max(cur_stage - 1, 0)
                    if cur_stage == 0:
                        break
                    cur_unit_ptx += 1
                    cond_latents = clean_latents_list[cur_stage][:, :, -(cur_unit_ptx * frame_per_unit) : -((cur_unit_ptx - 1) * frame_per_unit)]
                    stage_input.append(torch.cat([cond_latents] * 2) if do_classifier_free_guidance else cond_latents)

                if cur_stage == 0 and cur_unit_ptx < cur_unit_num:
                    cond_latents = clean_latents_list[0][:, :, :-(cur_unit_ptx * frame_per_unit)]
                    stage_input.append(torch.cat([cond_latents] * 2) if do_classifier_free_guidance else cond_latents)
            
                stage_input = list(reversed(stage_input))
                past_condition_latents.append(stage_input)

            intermed_latents = upscale_one_unit(
                latents[:,:,(unit_index - 1) * frame_per_unit:unit_index * frame_per_unit],
                past_condition_latents,
                prompt_embeds,
                prompt_attention_mask,
                pooled_prompt_embeds,
                num_inference_steps,
                height,
                width,
                frame_per_unit,
                device,
                dtype,
                generator=generator,
                is_first_frame=False,
            )

        generated_latents_list.append(intermed_latents[-1])

    generated_latents = torch.cat(generated_latents_list, dim=2)
    return generated_latents

def prepare_input_latents(control_latents, scheduler, device, dtype, generator=None):
    sigma = 1 - scheduler.end_sigmas[0]  # Signal coefficient at stage 0
    
    # Add noise to match the end of stage 0
    noise = torch.randn_like(control_latents)
    noised_latents = (1 - sigma) * noise + sigma * control_latents
    
    return noised_latents

with torch.autocast(device_type=device.type, dtype=dtype):
    latents = encode_video(low_res_frames).to(device=device, dtype=dtype)

    noised_latents = prepare_input_latents(
        latents,
        scheduler=scheduler,
        device=device, 
        dtype=dtype
    )

    generated_latents = upscale(noised_latents, prompt_embeds, prompt_attention_mask, pooled_prompt_embeds, num_inference_steps = [10, 10, 10])
```

**Comment by feifeiobama** [comment]
It would be different to train a flow matching model for super-resolution, see https://arxiv.org/abs/2312.07360.

We didn't notice this paper until recently, we find that Equations 9-10 of our paper somehow resemble "data-dependent coupling" in their Section 3.2.2.

**Comment by cyrildiagne** [comment]
Ok thanks a lot for the feedback and pointer, I'll check this paper.
